### ISSUE TITLE
Writing a DApp that buys/sells BNT

### ISSUE THREAD

**Issue by CurrencyTycoon** [bancorprotocol/contracts#26]
Hi,

I'm looking into writing a DApp that buys and sells BNT, for example a simple "options contract", or something along these lines.

Just playing around with it now.

I got the Buy BNT part working, by following the example in BancorBuyer.sol - simply calling the change() on the changer contract.

However, I cannot seem to do the reverse (selling).

Is there an example anywhere for how to do it?

I've tried something like this:

```
ISmartToken smartToken = tokenChanger.token();
         
         uint256 returnAmount = tokenChanger.change(smartToken, etherToken, getBNTBalance(msg.sender), 1); 

```

However, the function throws with "bad instruction" when i try it. 

https://etherscan.io/tx/0x5c6fda07aacfd5f39b6e55418c8244bb29f9cb74454ae0fe1fc5307e21889f8c

The `getBNTBalance` function gets the balance from the BNT Token contract in wei. For me, it returns `877843110001289470`

Am I missing something?

**Comment by rajpa** [comment]
You may have to convert etherToken to Ether...

**Comment by CurrencyTycoon** [comment]
@rajpa - yes, I understand that. However, my problem is changing BNT to etherToken from my smart contract, by calling changer contract.

**Comment by CurrencyTycoon** [comment]
btw, 877843110001289470 wei is 0.87784311000128947 bancor. Here is a gist of my contract that I have so far https://gist.github.com/CurrencyTycoon/1f490af9203b6e0891f1a7ab03fd9321

**Comment by yudilevi** [comment]
Hi!

@CurrencyTycoon I don't have the full code so I can't tell for sure but in order to execute the `change` function, the contract will have to have the caller's tokens in its own balance.

So the steps to do that would be -
- the user needs to 'approve' the contract to withdraw tokens from his/her account
- the contract transfers the tokens from the user to its own address
- the contract can then call `change` on the changer

**Comment by CurrencyTycoon** [comment]
Hi @yudilevi Thanks for your reply.

Despite your hints, I still cannot get the selling to work. However, I can get buying to work. Here is my buying code which works (taken from BancorBuyer.sol):

```

function long() payable returns (uint256 amount) {
        
        etherToken.deposit.value(msg.value)(); // deposit ETH in the reserve
        assert(etherToken.approve(tokenChanger, 0)); // need to reset the allowance to 0 before setting a new one
        assert(etherToken.approve(tokenChanger, msg.value)); // approve the changer to use the ETH amount for the purchase

        ISmartToken smartToken = tokenChanger.token();
        uint256 returnAmount = tokenChanger.change(etherToken, smartToken, msg.value, 1); // do the actual change using the current price
        assert(smartToken.transfer(msg.sender, returnAmount)); // transfer the tokens to the sender
        return returnAmount;
    
    }

```

So what I'm trying to do now is implement the missing sell() method form [BancorBoyer.sol](https://github.com/bancorprotocol/contracts/blob/master/solidity/contracts/BancorBuyer.sol) contract

I understand that when buying, we must approve the changer, and I can see that the `buy` method uses `transferFrom` which makes sense for me.

However, for the `sell()` in the changer, it appears that we do not need to call 'approve' before since a normal `transfer()` method is used internally by the [changer contract](https://github.com/bancorprotocol/contracts/blob/d8406c1ad614f6a9f6a68edd115a188ce1675411/solidity/contracts/BancorChanger.sol#L432)?

Here is what I have for the sell function:

(assuming that I have `getBNTBalance` method which gets the callers BNT ballance in wei. I've also initialized my tokenChanger in the constructor with  ITokenChanger and etherToken with IEtherToken)
```

So is the sell function I have so far (commented out the approve) is:

function sell() returns (uint256 returnAmount) {
        uint256 bal = getBNTBalance(msg.sender); // gets the balance of caller address in wei
        smartToken.transfer(this, bal); // transfer BNT tokens to this contract
        //assert(bnt.approve(tokenChanger, 0)); // need to reset the allowance to 0 before setting a new one
        //assert(bnt.approve(tokenChanger, bal));
        returnAmount = tokenChanger.change(smartToken, etherToken, bal, 1); // do the actual change using the current price
        assert(etherToken.transfer(msg.sender, returnAmount));
        return returnAmount;
         
     }

```

I'm not sure where it is throwing where I call it from my address that hold about 5 bnt: https://etherscan.io/tx/0x21d0b5ad43938bb2d84d1b8c27b98fc37a42884914bb4eff29d97ff6a1c6eab6

Any help with creating the sell function would be highly appreciated. Thanks.

**Comment by yudilevi** [comment]
Hi @CurrencyTycoon you can't call `transfer` on the SmartToken on behalf of another account - if that was possible, you could have simply emptied any account you wanted.
Instead, you'll need to use `approve` & `transferFrom`, there's currently no way around it.
The rest should work fine I think.

**Comment by CurrencyTycoon** [comment]
Hi @yudilevi - thanks a lot for your reply. It gave me some good hints and I came to the conclusion that I cannot do what I want to do, and will need to do a different approach.

I had a big "aha" moment when I realized that msg.sender and msg.value change depending on who the caller was, this means that I can't call `approve` from inside my contract, the user would need to call the BNT token contract directly!  I also cannot use `delegatecall`, since it doesn't allow me to access outside my storage, despite changing the msg.sender and msg.value to the original caller..

Anyhow, thanks again for your replies. Much appreciated.
